### PR TITLE
[CALCITE-4872] Make new UNKNOWN SqlTypeName, distinct from NULL (Will Noble)

### PR DESCRIPTION
**Pull request by wnob**
Make the unknown `SqlTypeName` and `RelDataType` distinct from the null `SqlTypeName` and `RelDataType`, and make the casting rules logical. See [CALCITE-4872](https://issues.apache.org/jira/browse/CALCITE-4872).

### CODE REVIEW

**Comment by zinking** [comment]
these changes look all pure theoretical, is there real cases , sqls regarding unknown type?

**Comment by wnob** [comment]
Yes, unfortunately Looker makes liberal use of `UNKNOWN`-typed expressions to represent arbitrary user-provided snippets of SQL. With a recent change making use of new rewrite rules, we ran into the issue of `UNKNOWN` types changing to `NULL` types when calling `createWithNullability`, and the rewrite rules failing.

File: core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java [hunk 1]
```diff
@@ -226,4 +227,29 @@ private void compareTypesIgnoringNullability(
...
+      BasicSqlType fromType;
+      try {
+        // This only works for basic types. Ignore the rest.
+        fromType = (BasicSqlType) f.typeFactory.createSqlType(fromTypeName);
+      } catch (AssertionError e) {
+        continue;
+      }
+      BasicSqlType nullableFromType = fromType.createWithNullability(!fromType.isNullable);
+
+      assertTrue(SqlTypeUtil.canCastFrom(unknownType, fromType, false));
```
[the remark this answers is not in the document]
Comment (vlsi) [reply core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:245]: I would rather suggest refactoring the test into `@ParameterizedTest` or creating a small `assertCanCastFrom(...)`, so the code becomes less noisy, and `assertCanCastFrom` could have clarification messages that explain which case failed (what was the input)

File: core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java [hunk 2]
```diff
@@ -54,7 +54,11 @@
    * @param typeName Type name
    */
   public BasicSqlType(RelDataTypeSystem typeSystem, SqlTypeName typeName) {
-    this(typeSystem, typeName, false, PRECISION_NOT_SPECIFIED,
```
[the remark this answers is not in the document]
Comment (wnob) [reply core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java:57]: I'm not 100% up to speed on the theoretical ramifications of this, but `BasicSqlType.createWithNullability` is already a thing (and it's existing usage in Calcite is essentially the motivation for this PR).

**Pull request closed without merging**